src/Adapter/FirehoseAdapter.php

```diff
@@ -0,0 +1,163 @@
+    protected $options;
+
+    /** @var string */
+    protected $deliveryStreamName;
+
+    /**
+     * @param FirehoseClient $client
+     * @param string    $deliveryStreamName
+     * @param array     $options
+     */
+    public function __construct(FirehoseClient $client, $deliveryStreamName, array $options = [])
+    {
+        $this->client = $client;
+        $this->deliveryStreamName = $deliveryStreamName;
+        $this->options = $options;
+    }
+
+    /**
+     * @param MessageInterface[] $messages
+     *
+     * @throws MethodNotSupportedException
+     */
+    public function acknowledge(array $messages)
+    {
+        throw new MethodNotSupportedException(
+            'acknowledge',
+            $this,
+            $messages
+        );
+    }
+
+    /**
+     * @param MessageFactoryInterface $factory
+     * @param int                     $limit
+     *
+     * @throws MethodNotSupportedException
+     */
+    public function dequeue(MessageFactoryInterface $factory, $limit)
+    {
+        throw new MethodNotSupportedException(
+            'dequeue',
+            $this,
+            []
+        );
+    }
+
+    /**
+     * @param MessageInterface[] $messages
+     *
+     * @throws FailedEnqueueException
+     */
+    public function enqueue(array $messages)
+    {
+        $failed = [];
+        $batches = array_chunk($this->createEnqueueEntries($messages), self::BATCHSIZE_SEND);
+
+        foreach ($batches as $batch) {
+            $requestRecords = array_map(function ($a) {
+                return [
+                    'Data' => json_encode($a)
+                ];
+            }, $batch);
+
+            $request = [
+                'DeliveryStreamName' => $this->deliveryStreamName,
+                'Records'  => $requestRecords,
+            ];
+
+            $results = $this->client->putRecordBatch($request);
+
+            foreach ($results->get('RequestResponses') as $idx => $response) {
+                if (isset($response['ErrorCode'])) {
+                    $failed[] = $messages[$batch[$idx]['Id']];
+                }
+            }
+        }
+
+        if (!empty($failed)) {
+            throw new FailedEnqueueException($this, $failed);
+        }
+    }
+
+    /**
+     * @throws MethodNotSupportedException
+     */
+    public function purge()
+    {
+        throw new MethodNotSupportedException(
+            'purge',
+            $this,
+            []
+        );
+    }
+
+    /**
+     * @throws MethodNotSupportedException
+     */
+    public function delete()
+    {
+        throw new MethodNotSupportedException(
+            'delete',
+            $this,
+            []
+        );
+    }
+
+    /**
+     * @param MessageInterface[] $messages
+     *
+     * @return array
+     */
+    protected function createEnqueueEntries(array $messages)
+    {
+        array_walk($messages, function (MessageInterface &$message, $id) {
+            $metadata = $message->getMetadata();
+            $message = [
```

Comment:
Is there a reason this has id, body and attributes ala sqs instead of just body?

Reply:
Nope, I thought that the expected behaviour was to pass the body and metadata through in the message, but release now that's just a specific SQS thing. I've simplified this to just pass through the body as the top-level message.